README.md
# React Native Switch Toggles 🔥

[![license](https://img.shields.io/github/license/mashape/apistatus.svg?style=for-the-badge)](https://github.com/nithinpp69/react-native-switch-toggles/blob/main/LICENSE)
![platforms](https://img.shields.io/badge/platforms-Android%20%7C%20iOS-brightgreen.svg?style=for-the-badge&colorB=191A17)
[![Version](https://img.shields.io/npm/v/react-native-switch-toggles.svg?style=for-the-badge)](https://www.npmjs.com/package/react-native-switch-toggles)
[![npm](https://img.shields.io/npm/dt/react-native-switch-toggles.svg?style=for-the-badge)](https://www.npmjs.com/package/react-native-switch-toggles)

A simple and customizable React Native switch component. 

## Demo

## ❤️ [Try on Expo Snack](https://snack.expo.dev/@nithinpp69/react-native-switch-toggles)

![](examples/demo.gif)
![](examples/demo2.gif)

## Prerequisites

 ⚠️ Peer Dependencies

 * [react-native-reanimated-v2](https://docs.swmansion.com/react-native-reanimated/docs/fundamentals/installation/)

This component has a peer dependency on react-native-reanimated-v2. react-native-reanimated-v2 has to be installed and linked into your project.
Follow [react-native-reanimated-v2](https://docs.swmansion.com/react-native-reanimated/docs/fundamentals/installation/) to install the dependency.

## Installation

 Supported version: react-native >= 0.59.0

  ```bash
  npm install react-native-switch-toggles
  ```
  
  or
  
  ```bash
  yarn add react-native-switch-toggles
  ```
  
## Example
```jsx
import Switch from 'react-native-switch-toggles';


const [isEnabled, setIsEnabled] = React.useState(false);
....

<Switch
  size={50}
  value={isEnabled}
  onChange={(value) => setIsEnabled(value)}
  renderOffIndicator={() => <Text style={{ fontSize: 16, color: 'white' }}>🌚</Text>}
  renderOnIndicator={() => <Text style={{ fontSize: 16, color: 'white' }}>🌝</Text>}
/>

<Switch
  size={50}
  value={isEnabled}
  onChange={(value) => setIsEnabled(value)}
  activeTrackColor={"#45D058"}
  renderOffIndicator={() => <Text style={{ fontSize: 16, color: 'white' }}>OFF</Text>}
  renderOnIndicator={() => <Text style={{ fontSize: 16, color: 'white' }}>ON</Text>}
/>

```
![](examples/demo.gif)

```jsx
import Switch from 'react-native-switch-toggles';


const [isEnabled, setIsEnabled] = React.useState(false);
....

<>
  <Text style={styles.label}>Simple Switch</Text>
  <Switch
    value={isEnabled}
    onChange={(value) => setIsEnabled(value)}
  />
</>


<>
  <Text style={styles.label}>Switch with on/off indicator</Text>
  <Switch
    value={isEnabled}
    onChange={(value) => setIsEnabled(value)}
    activeTrackColor={'#45D058'}
    renderOffIndicator={() => (
      <Text style={{ fontSize: 8, color: 'white' }}>OFF</Text>
    )}
    renderOnIndicator={() => (
      <Text style={{ fontSize: 8, color: 'white' }}>ON</Text>
    )}
  />
</>


<>
  <Text style={styles.label}>Switch with on/off thumb indicator</Text>
  <Switch
    size={40}
    value={isEnabled}
    onChange={(value) => setIsEnabled(value)}
    activeTrackColor={'#6ab04c'}
    renderInactiveThumbIcon={() => (
      <Text style={{ fontSize: 12, color: 'black' }}>💩</Text>
    )}
    renderActiveThumbIcon={() => (
      <Text style={{ fontSize: 12, color: 'black' }}>🔥</Text>
    )}
  />
</>


<>
  <Text style={styles.label}>
    Switch with on/off thumb/track indicator
  </Text>
  <Switch
    size={40}
    value={isEnabled}
    onChange={(value) => setIsEnabled(value)}
    activeTrackColor={'#6ab04c'}
    inactiveTrackColor={'#eb4d4b'}
    renderInactiveThumbIcon={() => (
      <Text style={{ fontSize: 12, color: 'black' }}>💩</Text>
    )}
    renderActiveThumbIcon={() => (
      <Text style={{ fontSize: 12, color: 'black' }}>🔥</Text>
    )}
    renderOffIndicator={() => (
      <Text style={{ fontSize: 12, color: 'white' }}>OFF</Text>
    )}
    renderOnIndicator={() => (
      <Text style={{ fontSize: 12, color: 'white' }}>ON</Text>
    )}
  />
</>


<>
  <Text style={styles.label}>Big switch</Text>
  <Switch
    size={60}
    value={isEnabled}
    onChange={(value) => setIsEnabled(value)}
    activeTrackColor={'#4b7bec'}
    inactiveTrackColor={'#ff7f50'}
    renderInactiveThumbIcon={() => (
      <Text style={{ fontSize: 16, color: 'black' }}>💩</Text>
    )}
    renderActiveThumbIcon={() => (
      <Text style={{ fontSize: 16, color: 'black' }}>🔥</Text>
    )}
    renderOffIndicator={() => (
      <Text style={{ fontSize: 12, color: 'white' }}>OFF</Text>
    )}
    renderOnIndicator={() => (
      <Text style={{ fontSize: 12, color: 'white' }}>ON</Text>
    )}
  />
</>


<>
  <Text style={styles.label}>Big switch with thumb color</Text>
  <Switch
    size={60}
    value={isEnabled}
    onChange={(value) => setIsEnabled(value)}
    activeThumbColor={'#f9ca24'}
    inactiveThumbColor={'#6ab04c'}
    activeTrackColor={'#6ab04c'}
    inactiveTrackColor={'#ffffff'}
    renderInactiveThumbIcon={() => (
      <Text style={{ fontSize: 14, color: 'black' }}>💩</Text>
    )}
    renderActiveThumbIcon={() => (
      <Text style={{ fontSize: 14, color: 'black' }}>🔥</Text>
    )}
    renderOffIndicator={() => (
      <Text style={{ fontSize: 12, color: 'black' }}>OFF</Text>
    )}
    renderOnIndicator={() => (
      <Text style={{ fontSize: 12, color: 'white' }}>ON</Text>
    )}
  />
</>

```
![](examples/demo2.gif)

## Props
| Prop                        | Description                                                                           | Type                          | Default Value              | Required  |
| :--------------------------:|:--------------------------------------------------------------------------------------|:-----------------------------:|:--------------------------:|:---------:|
| value                       | switch on/off state value                                                             | Boolean                       |                            | true      |
| onChange                    | callback on switch value change                                                       | Function                      |  (value: boolean) => void; | true      |
| size                        | size of the switch component                                                          | Number                        |  25                        | false     |
| disabled                    | enable/disable switch                                                                 | Boolean                       |  false                     | false     |
| activeTrackColor            | track color when switch value is true                                                 | String                        |  "rgba(255,255,255,0.6)" | false     |
| inactiveTrackColor          | track color when switch value is false                                                | String                        |  "rgba(0,0,0,0.2)"       | false     |
| activeThumbColor            | thumb color when switch value is true                                                 | String                        |  "#ffffff"               | false     |
| inactiveThumbColor          | thumb color when switch value is false                                                | String                        |  "#ffffff"               | false     |
| renderOnIndicator           | render a custom view on switch track when the switch value is true                    | Function                      |  () => null                | false     |
| renderOffIndicator          | render a custom view on switch track when the switch value is false                   | Function                      |  () => null                | false     |
| renderActiveThumbIcon       | render a custom view on switch thumb when the switch value is true                    | Function                      |  () => null                | false     |
| renderInactiveThumbIcon     | render a custom view on switch thumb when the switch value is false                   | Function                      |  () => null                | false     |

## License
This project is licenced under the MIT License.


CHANGELOG.md
###  Build: 🏠 `1.0.1` - react-native-switch-toggles

---
- fix: fix android shadow issue

###  Build: 🏠 `1.0.0` - react-native-switch-toggles

---
- chore: added build setup with react-native-builder-bob.
- chore: publish compiled source to npm


src/switch/types.ts
import React from 'react';

interface SwitchProps {
  /**
   * size of the switch. Use this prop to set the size of the
   * switch.
   * 
   * @default 25
   */
  size?: number;
  /**
   * switch on/off state value. Pass this prop to set the initial
   * state of the switch.
   */
  value: boolean;
  /**
   * enable/disable switch. Use this prop to enable/disable the
   * switch.
   * 
   * @default false
   */
  disabled?: boolean;
  /**
   * callback on switch value change. This is useful if you want to
   * run a function when the switch value changes.
   * 
   * @param value - switch value
   */
  onChange: (value: boolean) => void;
  /**
   * track color when switch value is true. Use this to customize the
   * switch track color when switch value is true.
   * 
   * @default 'rgba(255,255,255,0.6)'
   */
  activeTrackColor?: string;
  /**
   * track color when switch value is false. Use this to customize the
   * switch track color when switch value is false.
   * 
   * @default 'rgba(0,0,0,0.2)
   */
  inactiveTrackColor?: string;
  /**
   * thumb color when switch value is true. Use this to customize the
   * switch thumb color when switch value is true.
   * 
   * @default '#ffffff
   */
  activeThumbColor?: string;
  /**
   * thumb color when switch value is false. Use this to customize the
   * switch thumb color when switch value is false.
   * 
   * @default '#ffffff
   */
  inactiveThumbColor?: string;
  /**
   * render a custom view on switch track when the switch value is true.
   */
  renderOnIndicator?: () => React.ReactNode | null;
  /**
   * render a custom view on switch track when the switch value is false
   */
  renderOffIndicator?: () => React.ReactNode | null;
  /**
   * ender a custom view on switch thumb when the switch value is true
   */
  renderActiveThumbIcon?: () => React.ReactNode | null;
  /**
   * render a custom view on switch thumb when the switch value is false
   */
  renderInactiveThumbIcon?: () => React.ReactNode | null;
}

export type {
  SwitchProps,
};


src/switch/styles.ts
import {StyleSheet} from 'react-native';

import Colors from '../utils/colors';

export type StyleProps = {
  switchWidth: number,
  switchHeight: number,
  switchBorderRadius: number,
  disabled: boolean,
};

const styles = (props: StyleProps) =>
  StyleSheet.create({
    switchContainer: {
      opacity: props.disabled ? 0.6 : 1,
    },
    container: {
      width: props.switchWidth,
      height: props.switchHeight,
      backgroundColor: 'rgba(0,0,0,0.25)',
      borderRadius: props.switchBorderRadius,
      justifyContent: 'center',
      overflow: 'hidden',
    },
    thumb: {
      height: props.switchHeight - 10,
      width: props.switchHeight - 10,
      borderRadius: props.switchBorderRadius,
      margin: 5,
      backgroundColor: Colors.white,
      shadowColor: '#000',
      shadowOffset: {
        width: 0,
        height: 2,
      },
      shadowOpacity: 0.25,
      shadowRadius: 3.84,
      elevation: 5,
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
      overflow: 'hidden',
    },
    onIndicator: {
      width: '50%',
      height: '100%',
      backgroundColor: 'transparent',
      position: 'absolute',
      left: 0,
      alignItems: 'center',
      justifyContent: 'center',
    },
    offIndicator: {
      width: '50%',
      height: '100%',
      backgroundColor: 'transparent',
      position: 'absolute',
      right: 0,
      alignItems: 'center',
      justifyContent: 'center',
    },
    thumbIconContainer: {
      flex: 1,
      alignItems: 'center',
      justifyContent: 'center',
    },
  });

export default styles;


src/switch/types.d.ts
import React from "react";

export interface SwitchProps {
    /**
     * size of the switch
     */
    size?: number;
    /**
     * switch on/off state value
     */
    value: boolean;
    /**
     * enable/disable switch
     */
    disabled?: boolean;
    /**
     * callback on switch value change
     */
    onChange: (value: boolean) => void;
    /**
     * track color when switch value is true
     */
    activeTrackColor?: string;
    /**
     * track color when switch value is false
     */
    inactiveTrackColor?: string;
    /**
     * thumb color when switch value is true
     */
    activeThumbColor?: string;
    /**
     * thumb color when switch value is false
     */
    inactiveThumbColor?: string;
    /**
     * render a custom view on switch track when the switch value is true
     */
    renderOnIndicator?: () => React.ReactNode;
    /**
     * render a custom view on switch track when the switch value is false
     */
    renderOffIndicator?: () => React.ReactNode;
    /**
     * ender a custom view on switch thumb when the switch value is true
     */
    renderActiveThumbIcon?: () => React.ReactNode;
    /**
     * render a custom view on switch thumb when the switch value is false
     */
    renderInactiveThumbIcon?: () => React.ReactNode;
}

declare const Switch: React.FC<SwitchProps>;

export default Switch;


src/types/index.ts
import type {ReactNode} from 'react';

interface SwitchProps {
  /**
   * size of the switch. Use this prop to set the size of the
   * switch.
   *
   * @default 25
   */
  size?: number;
  /**
   * switch on/off state value. Pass this prop to set the initial
   * state of the switch.
   */
  value: boolean;
  /**
   * enable/disable switch. Use this prop to enable/disable the
   * switch.
   *
   * @default false
   */
  disabled?: boolean;
  /**
   * callback on switch value change. This is useful if you want to
   * run a function when the switch value changes.
   *
   * @param value - switch value
   */
  // eslint-disable-next-line no-unused-vars
  onChange: (value: boolean) => void;
  /**
   * track color when switch value is true. Use this to customize the
   * switch track color when switch value is true.
   *
   * @default 'rgba(255,255,255,0.6)'
   */
  activeTrackColor?: string;
  /**
   * track color when switch value is false. Use this to customize the
   * switch track color when switch value is false.
   *
   * @default 'rgba(0,0,0,0.2)
   */
  inactiveTrackColor?: string;
  /**
   * thumb color when switch value is true. Use this to customize the
   * switch thumb color when switch value is true.
   *
   * @default '#ffffff
   */
  activeThumbColor?: string;
  /**
   * thumb color when switch value is false. Use this to customize the
   * switch thumb color when switch value is false.
   *
   * @default '#ffffff
   */
  inactiveThumbColor?: string;
  /**
   * render a custom view on switch track when the switch value is true.
   */
  renderOnIndicator?: () => ReactNode | null;
  /**
   * render a custom view on switch track when the switch value is false
   */
  renderOffIndicator?: () => ReactNode | null;
  /**
   * ender a custom view on switch thumb when the switch value is true
   */
  renderActiveThumbIcon?: () => ReactNode | null;
  /**
   * render a custom view on switch thumb when the switch value is false
   */
  renderInactiveThumbIcon?: () => ReactNode | null;
}

export type {SwitchProps};
